rasa/nlu/featurizers/sparse_featurizer/regex_featurizer.py

```diff
@@ -42,31 +44,133 @@ def required_components(cls) -> List[Type[Component]]:
+                f" was left to `None`. This is invalid since the featurizer"
+                f" needs vocabulary statistics to featurize in finetune mode."
+            )
+
+    @lazy_property
+    def vocabulary_stats(self) -> Dict[Text, int]:
+        """Compute total vocabulary size and how much of it is consumed.
+
+        Returns:
+            Computed vocabulary size and number of filled vocabulary slots.
+        """
+        if not self.finetune_mode:
+            max_number_patterns = (
+                len(self.known_patterns) + self._get_num_additional_slots()
+            )
+            return {
+                "pattern_slots_filled": len(self.known_patterns),
+                "max_number_patterns": max_number_patterns,
+            }
+        else:
+            self.pattern_vocabulary_stats["pattern_slots_filled"] = len(
+                self.known_patterns
+            )
+            return self.pattern_vocabulary_stats
+
+    def _merge_new_patterns(self, new_patterns: List[Dict[Text, Text]]):
+        """Update already known patterns with new patterns extracted from data.
+
+        Args:
+            new_patterns: Patterns extracted from training data and to be merged with known patterns.
+        """
+        max_number_patterns = self.pattern_vocabulary_stats["max_number_patterns"]
+        pattern_name_index_map = {
+            pattern["name"]: index for index, pattern in enumerate(self.known_patterns)
+        }
+        patterns_dropped = False
+
+        for extra_pattern in new_patterns:
+            new_pattern_name = extra_pattern["name"]
+
+            # Some patterns may have just new examples added
+            # to them. These do not count as additional pattern.
+            if new_pattern_name in pattern_name_index_map:
+                self.known_patterns[pattern_name_index_map[new_pattern_name]][
+                    "pattern"
+                ] = extra_pattern["pattern"]
+            else:
+                if len(self.known_patterns) == max_number_patterns:
+                    patterns_dropped = True
+                    continue
+                self.known_patterns.append(extra_pattern)
+        if patterns_dropped:
+            logger.warning(
+                f"The originally trained model was configured to "
+                f"handle a maximum number of {max_number_patterns} patterns. "
+                f"The current training data exceeds this number as "
+                f"there are {len(new_patterns)} patterns in total. "
+                f"Some patterns will be dropped and not used for "
+                f"featurization. It is advisable to re-train the "
+                f"model from scratch."
+            )
+
+    def _get_num_additional_slots(self) -> int:
+        """Compute number of additional pattern slots available in vocabulary on top of known patterns."""
+        if self.number_additional_patterns is None:
+            self.number_additional_patterns = max(10, len(self.known_patterns) * 2)
```

Comment:
Why `max(10, len(self.known_patterns) * 2)`? Can you please add a comment and maybe use constants if it makes sense.

Reply:
Explained in a comment and added a constant for `10`.